util/src/main/java/com/coderanch/util/cli/InputUtility.java

```diff
@@ -0,0 +1,264 @@
+     *
+     * @param prompt          the prompt to display to the user.
+     * @param doublePredicate the predicate to use for validation.
+     * @return a validated floating point number.
+     * @throws IOException if there's a problem while reading from the underlying stream.
+     */
+    public double nextDouble(String prompt, Predicate<? super Double> doublePredicate) throws IOException {
+        System.out.println(prompt);
+        String line;
+        OptionalDouble num;
+        for (
+            line = reader.readLine(), num = OptionalDouble.empty();
+            num.isEmpty();
+            line = reader.readLine()
+        ) {
+            num = tryDoubleParse(line, doublePredicate);
+            if (num.isEmpty()) {
+                System.out.println("Invalid input.");
+            }
+        }
+        return num.getAsDouble();
+    }
+
+    /**
+     * Tries to parse a Double from the string and validate it.
+     *
+     * @param line            the string to parse.
+     * @param doublePredicate the predicate used to validate the double.
+     * @return a validated Double or nothing.
+     */
+    private OptionalDouble tryDoubleParse(String line, Predicate<? super Double> doublePredicate) {
+        try {
+            var num = Double.parseDouble(line);
+            return doublePredicate.test(num) ? OptionalDouble.of(num) : OptionalDouble.empty();
+        }
+        catch (NumberFormatException e) {
+            return OptionalDouble.empty();
+        }
+    }
+
+    /**
+     * Gets the next yes or no answer.
+     * Calls {@link InputUtility#nextString(String, Predicate)}.
+     * Passes the prompt and a predicate that validates "y", "n", "yes", "no".
+     *
+     * @param prompt the prompt to display to the user.
+     * @return true for yes and false for no.
+     * @throws IOException if there's a problem while reading from the underlying stream.
+     */
+    public boolean nextYesNo(String prompt) throws IOException {
+        var result = this.nextString(prompt, yesOrNo()).trim().toLowerCase();
+        return YES_SYNONYMS.contains(result);
+    }
+
+    /**
+     * Waits for the user to press &lt;enter&gt;.
+     * The prompt defaults to "Press when ready".
+     * Calls {@link InputUtility#pause(String)}.
+     *
+     * @throws IOException if there's a problem while reading from the underlying stream.
+     */
+    public void pause() throws IOException {
+        pause("Press enter when ready");
+    }
+
+    /**
+     * Uses prompt entered and waits for the user to press &lt;enter&gt;.
+     * Calls nextString{@link InputUtility#nextString(String, Predicate)}.
+     *
+     * @param prompt the prompt to display to the user.
+     * @throws IOException if there's a problem while reading from the underlying stream.
+     */
+    public void pause(String prompt) throws IOException {
+        this.nextString(prompt, String::isBlank);
```

Comment:
I wouldn't force a blank line. If there's garbage input on the line, that doesn't mean that the user didn't press enter, which is all we care about.

Normally I'd say, call a method overloads that doesn't accept a predicate and is just always happy, but this PR is getting kinda long so we can fix that in the future.

For now, just pass a predicate that returns true.